transport/grpc.go

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import (
+	"bytes"
+	"errors"
+	"io/ioutil"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/multierr"
+	apipeer "go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/peer"
+	"go.uber.org/yarpc/peer/hostport"
+	"go.uber.org/yarpc/peer/roundrobin"
+	"go.uber.org/yarpc/transport/grpc"
+	"golang.org/x/net/context"
+)
+
+var (
+	errGRPCNoAddresses = errors.New("must specify at least one grpc address")
+	errGRPCNoTracer    = errors.New("must specify grpc tracer")
+	errGRPCNoCaller    = errors.New("must specify grpc caller")
+	errGRPCNoService   = errors.New("must specify grpc service")
+	errGRPCNoEncoding  = errors.New("must specify grpc encoding")
+	errGRPCNoProcedure = errors.New("must specify grpc procedure")
+)
+
+// GRPCOptions are used to create a GRPC transport.
+type GRPCOptions struct {
+	Addresses       []string
+	Tracer          opentracing.Tracer
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+}
+
+// NewGRPC returns a transport that calls a GRPC service.
+func NewGRPC(options GRPCOptions) (TransportCloser, error) {
+	return newGRPC(options)
+}
+
+type grpcTransport struct {
+	Transport       transport.Transport
+	Outbound        transport.UnaryOutbound
+	Caller          string
+	Encoding        string
+	RoutingKey      string
+	RoutingDelegate string
+	tracer          opentracing.Tracer
+}
+
+func newGRPC(options GRPCOptions) (*grpcTransport, error) {
+	if len(options.Addresses) == 0 {
+		return nil, errGRPCNoAddresses
+	}
+	if options.Tracer == nil {
+		return nil, errGRPCNoTracer
+	}
+	if options.Caller == "" {
+		return nil, errGRPCNoCaller
+	}
+	if options.Encoding == "" {
+		return nil, errGRPCNoEncoding
+	}
+
+	identifiers := make([]apipeer.Identifier, len(options.Addresses))
```

Comment:
can you make a private function, it's a pure function that does `[]string -> []apipeer.Identifier`

Reply:
Done